packages/next/pages/_document.tsx

```diff
@@ -243,27 +243,55 @@ export class Head extends Component<
   }
 
   getPreloadMainLinks(files: DocumentFiles): JSX.Element[] | null {
-    const { assetPrefix, devOnlyCacheBusterQueryString } = this.context
+    const {
+      assetPrefix,
+      devOnlyCacheBusterQueryString,
+      scriptLoader,
+    } = this.context
     const preloadFiles = files.allFiles.filter((file: string) => {
       return file.endsWith('.js')
     })
 
-    return !preloadFiles.length
-      ? null
-      : preloadFiles.map((file: string) => (
-          <link
-            key={file}
-            nonce={this.props.nonce}
-            rel="preload"
-            href={`${assetPrefix}/_next/${encodeURI(
-              file
-            )}${devOnlyCacheBusterQueryString}`}
-            as="script"
-            crossOrigin={
-              this.props.crossOrigin || process.env.__NEXT_CROSS_ORIGIN
-            }
-          />
-        ))
+    return [
+      ...(scriptLoader.eager || []).map((file) => (
+        <link
+          key={file.src}
+          nonce={this.props.nonce}
+          rel="preload"
+          href={file.src}
+          as="script"
+          crossOrigin={
+            this.props.crossOrigin || process.env.__NEXT_CROSS_ORIGIN
+          }
+        />
+      )),
+      ...preloadFiles.map((file: string) => (
+        <link
+          key={file}
+          nonce={this.props.nonce}
+          rel="preload"
+          href={`${assetPrefix}/_next/${encodeURI(
+            file
+          )}${devOnlyCacheBusterQueryString}`}
+          as="script"
+          crossOrigin={
+            this.props.crossOrigin || process.env.__NEXT_CROSS_ORIGIN
+          }
+        />
+      )),
+      ...(scriptLoader.defer || []).map((file: string) => (
+        <link
+          key={file}
+          nonce={this.props.nonce}
+          rel="preload"
+          href={file}
+          as="script"
+          crossOrigin={
+            this.props.crossOrigin || process.env.__NEXT_CROSS_ORIGIN
+          }
+        />
+      )),
```

Comment:
Is the difference between `lazy` and `defer` only that defer'd scripts get preloaded? Is this a large enough difference to justify the existence of two modes? Can you please explain a scenario where `lazy` wouldn't be an appropriate replacement for `defer`?

Reply:
`lazy` is after onload - basically we wait for all resources to be downloaded
`defer` is on hydration - here were start loading after prioritizing only 1P scripts

This difference may seem subtle but we've seen traces of apps where the onload can be significantly delayed because of cascading requests. So onload and onHydration can have a large time delta in terms of when they trigger a script load.

The decision is also nuanced depending on the type of app, what resources they prioritize, what type of 3P scripts they load etc. **So we wanted more options initially to experiment and then determine the right defaults.**

A potential example of using `lazy` vs `defer` is we were experimenting with loading GTM with `lazy` and something like Social login/video player with `defer` (because onHydration the buttons become interactive and you don't want to wait for images to load)